x/operator/keeper/state_update.go

```diff
@@ -0,0 +1,494 @@
+		SlashContract: slashContract,
+		// #nosec G701
+		OptedInHeight:  uint64(ctx.BlockHeight()),
+		OptedOutHeight: types.DefaultOptedOutHeight,
+	}
+	err = k.UpdateOptedInfo(ctx, operatorAddress.String(), avsAddr, optedInfo)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// OptOut call this function to opt out of AVS
+func (k *Keeper) OptOut(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr string) error {
+	// check optedIn info
+	if !k.IsOptedIn(ctx, operatorAddress.String(), avsAddr) {
+		return types.ErrNotOptedIn
+	}
+
+	// get the assets supported by the AVS
+	avsSupportedAssets, err := k.avsKeeper.GetAvsSupportedAssets(ctx, avsAddr)
+	if err != nil {
+		return err
+	}
+	// get the Assets opted in the operator
+	operatorAssets, err := k.restakingStateKeeper.GetOperatorAssetInfos(ctx, operatorAddress, avsSupportedAssets)
+	if err != nil {
+		return err
+	}
+
+	assetFilter := make(map[string]interface{})
+
+	for assetID := range operatorAssets {
+		err = k.DeleteAssetState(ctx, assetID, avsAddr, operatorAddress.String())
+		if err != nil {
+			return err
+		}
+		assetFilter[assetID] = nil
+	}
+
+	avsOperatorTotalValue, err := k.GetOperatorShare(ctx, avsAddr, operatorAddress.String())
+	if err != nil {
+		return err
+	}
+	if avsOperatorTotalValue.IsNegative() {
+		return errorsmod.Wrap(types.ErrTheValueIsNegative, fmt.Sprintf("OptOut,avsOperatorTotalValue:%suite", avsOperatorTotalValue))
+	}
+
+	// delete the share value of operator itself, the input stakerID should be empty
+	err = k.DeleteStakerShare(ctx, avsAddr, "", operatorAddress.String())
+	if err != nil {
+		return err
+	}
+
+	// UpdateAVSShare
+	err = k.UpdateAVSShare(ctx, avsAddr, avsOperatorTotalValue.Neg())
+	if err != nil {
+		return err
+	}
+	// DeleteOperatorShare
+	err = k.DeleteOperatorShare(ctx, avsAddr, operatorAddress.String())
+	if err != nil {
+		return err
+	}
+
+	// DeleteStakerShare
+	relatedAssetsState, err := k.delegationKeeper.DelegationStateByOperatorAssets(ctx, operatorAddress.String(), assetFilter)
+	if err != nil {
+		return err
+	}
+	for stakerID := range relatedAssetsState {
+		err = k.DeleteStakerShare(ctx, avsAddr, stakerID, operatorAddress.String())
+		if err != nil {
+			return err
+		}
+	}
+
+	// set opted-out height
+	optedInfo, err := k.GetOptedInfo(ctx, operatorAddress.String(), avsAddr)
+	if err != nil {
+		return err
+	}
+	// #nosec G701
+	optedInfo.OptedOutHeight = uint64(ctx.BlockHeight())
+	err = k.UpdateOptedInfo(ctx, operatorAddress.String(), avsAddr, optedInfo)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetAssetsAmountToSlash It will slash the assets that are opting into AVS first, and if there isn't enough to slash, then it will slash the assets that have requested to undelegate but still locked.
+func (k *Keeper) GetAssetsAmountToSlash(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr string, occurredSateHeight int64, slashProportion sdkmath.LegacyDec) (*SlashAssets, error) {
```

Comment:
What is `occurredSateHeight` ? Looks like a typo.

Reply:
It's the height at which the slash event occurs, I will rename it to `slashEventHeight`.